### PR TITLE
Count processors via affinity mask on linux

### LINKED ISSUE #12: Doesn't take CPU affinity into account
While `num_cpus()` gives the correct number of physical CPUs, it does not take into account which of those CPU's the current process is actually allowed to use. It would be neat to be able to specify which of the two numbers to get.

For example, consider a simple `num_cpus` program that just prints the value of `num_cpus::get();`, and then exits:

```
$ nproc
80
$ num_cpus
80
$ taskset -c 0,1 nproc
2
$ taskset -c 0,1 num_cpus
80
```

If it's at all helpful, the source for `nproc` (which is included in coreutils) can be seen [here](https://fossies.org/dox/coreutils-8.23/lib_2nproc_8c_source.html).

### PR DESCRIPTION
**Pull request by kubo39**
This Fixes #12

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -146,16 +146,67 @@ fn get_num_cpus() -> usize {
...
+    c = (c & 0x3333333333333333) + ((c >> 2) & 0x3333333333333333);
+    c = (c & 0x0f0f0f0f0f0f0f0f) + ((c >> 4) & 0x0f0f0f0f0f0f0f0f);
+    c = (c & 0x00ff00ff00ff00ff) + ((c >> 8) & 0x00ff00ff00ff00ff);
+    c = (c & 0x0000ffff0000ffff) + ((c >> 16) & 0x0000ffff0000ffff);
+    c = (c & 0x00000000ffffffff) + ((c >> 32) & 0x00000000ffffffff);
+    c
+}
+
+#[cfg(target_os = "linux")]
+fn get_num_cpus() -> usize {
```
Comment (seanmonstar) [inline src/lib.rs]: I wonder whether using affinity by default is better, or if it'd be a sort of breaking change. Would one always always want the affinity, or could one conceivably want to just know the number or CPUs on the machine anyways?

Put another way, should this be a separate function, or just be the default and only implementation for linux? I'm not saying one or the other is correct, I'd be interested in rationale for either case.

File: src/lib.rs [hunk 2]
```diff
@@ -146,16 +146,67 @@ fn get_num_cpus() -> usize {
...
-        target_os = "linux",
-        target_os = "nacl",
-        target_os = "macos",
-        target_os = "ios",
-        target_os = "android",
-        target_os = "solaris",
-        target_os = "fuchsia",
-    )
+#[cfg(feature = "nightly")]
+#[feature(asm, naked_functions)]
```
Comment (seanmonstar) [inline src/lib.rs]: Is there large benefits to using this asm as opposed to not requiring unstable features and using the bit masking?

Comment (sfackler) [inline src/lib.rs]: This function exists in the standard library: https://doc.rust-lang.org/std/primitive.u64.html#method.count_ones

Comment (kubo39) [inline src/lib.rs]: It's just taken from my perfromance sensitive code written in C.
`count_ones` is just what I wanted, thanks!

**Comment by seanmonstar** [comment]
I've moved the question out of the line comment, so that edits don't make it disappear:



>I wonder whether using affinity by default is better, or if it'd be a sort of breaking change. Would one always always want the affinity, or could one conceivably want to just know the number or CPUs on the machine anyways?
>
>Put another way, should this be a separate function, or just be the default and only implementation for linux? I'm not saying one or the other is correct, I'd be interested in rationale for either case.

**Comment by kubo39** [comment]
I researched some other language implementations on Linux.
golang and nproc(1) counts cores via affinity mask, python/Ruby(parallel gem)/D does not.
ref: https://github.com/kubo39/cpucount

**Comment by seanmonstar** [comment]
I'm slightly inclined to follow golang, but welcome other opinions. I also wonder if that's breaking change (or a "bug fix").

**Comment by seanmonstar** [comment]
My current feelings are:

- `get()` should check affinity. It's the "default" function to call, and I assume most people using it are just wanting to know how many threads or processes they should spawn, so it's actually bad to report more CPUs than the process can actually use.
- `get_physical()` is already there for those who need that specific datum.
- `get_all()` or something could be added later if there is really a need to know ALL cores, physical and virtual.

So, I'm considering merging this as-is, unless there's other thoughts.

**Comment by brson** [comment]
@seanmonstar I agree that the default behavior should give you a reasonable number of threads to spawn.

I'm not sure about whether it's a breaking change or not, but I'd say it should get a major bump.

**Comment by pedrocr** [comment]
@seanmonstar the reason for `get_physical()` is to get the same thing as `get()` but without hyperthreading. I agree that it makes sense that `get()` returns the threads you can use but then `get_physical()` should return the physical threads you can use. Otherwise it's inconsistent. So I think it makes more sense to do 4 functions having normal/physical versions of with/without affinity and have the default be the affinity ones. So:

  * `get()` is all threads with affinity
  * `get_physical()` is all physically sepparate threads with affinity
  * `get_all()` is all threads without affinity
  * `get_all_physical()` is the `current get_physical()`

To do this what we're missing is the `get_physical()` implementation where affinity is taken into account.

**Comment by seanmonstar** [comment]
@pedrocr that makes sense, thanks for the insight! I've opened #40 for tracking that.

**Comment by seanmonstar** [comment]
Thanks @kubo39! Excellent work!

**Pull request merged**